src/Components/Components/src/NavigationManager.cs

```diff
@@ -274,6 +281,128 @@ protected void NotifyLocationChanged(bool isInterceptedLink)
+        }
+        finally
+        {
+            ArrayPool<IHandleLocationChanging>.Shared.Return(locationChangingHandlersCopy);
+        }
+    }
+
+    /// <summary>
+    /// Sets whether there are active location changing handlers.
+    /// </summary>
+    /// <param name="value">Whether there are active location changing handlers.</param>
+    /// <returns><see langword="true"/> if the change was applied successfully, otherwise <see langword="false"/>.</returns>
+    protected virtual bool SetHasLocationChangingHandlers(bool value)
+        => true;
+
+    /// <summary>
+    /// Invokes <see cref="SetHasLocationChangingHandlers(bool)"/> if the existence of location changing handlers has changed.
+    /// </summary>
+    protected void UpdateHasLocationChangingHandlers()
+    {
+        var hasLocationChangingHandlers = _locationChangingHandlers.Count != 0;
+        if (hasLocationChangingHandlers != _lastUpdatedHasLocationChangingHandlers)
+        {
+            if (SetHasLocationChangingHandlers(hasLocationChangingHandlers))
+            {
+                _lastUpdatedHasLocationChangingHandlers = hasLocationChangingHandlers;
+            }
+        }
+    }
+
+    /// <summary>
+    /// Adds a handler to process incoming navigation events.
+    /// </summary>
+    /// <param name="locationChangingHandler">The handler to process incoming navigation events.</param>
+    public void AddLocationChangingHandler(IHandleLocationChanging locationChangingHandler)
+    {
+        AssertInitialized();
+        _locationChangingHandlers.Add(locationChangingHandler);
+        UpdateHasLocationChangingHandlers();
+    }
+
+    /// <summary>
+    /// Removes a previously-added location changing handler.
+    /// </summary>
+    /// <param name="locationChangingHandler">The handler to remove.</param>
+    /// <returns><see langword="true"/> if the handler was removed, otherwise <see langword="false"/>.</returns>
+    public bool RemoveLocationChangingHandler(IHandleLocationChanging locationChangingHandler)
+    {
+        AssertInitialized();
+
+        if (_locationChangingHandlers.Remove(locationChangingHandler))
+        {
+            UpdateHasLocationChangingHandlers();
+            return true;
+        }
+
+        return false;
+    }
+
+    /// <summary>
+    /// Enables a navigation prompt to be displayed before navigation events complete.
+    /// </summary>
+    /// <param name="message">The message to display in the prompt. Note that some browsers will ignore this message for external navigations.</param>
+    /// <param name="externalNavigationsOnly">If <see langword="true"/>, the prompt will not display for internal page navigations.</param>
+    /// <returns>A <see cref="ValueTask"/> representing the completion of the operation.</returns>
+    protected internal virtual ValueTask EnableNavigationPromptAsync(string message, bool externalNavigationsOnly)
```

Comment:
Are EnableNavigationPromptAsync/DisableNavigationPromptAsync really needed inside NavigationManager? I would have thought that the new `NavigationPrompt` component could register its own `beforeunload` handler via JS interop without involving the core NavigationManager.

Most likely I'm missing something, but it would be great to clarify. Thanks!

Reply:
Since `NavigationPrompt` lives in `Microsoft.AspNetCore.Components.Routing`, using an `IJSRuntime` in the component would mean adding a reference to `Microsoft.JSInterop` in `Microsoft.AspNetCore.Components`. So, I added these APIs to `NavigationManager` to move the JS interop code out of `M.A.Components`.

An alternative could be moving `NavigationPrompt` to `Microsoft.AspNetCore.Components.Web`. What do you think?

Reply:
> An alternative could be moving NavigationPrompt to Microsoft.AspNetCore.Components.Web. What do you think?

Yep, that sounds great!

The high-level functionality here really is quite web-specific, particularly in the distinction between internal and external navigation (which we don't have any reason to think would have a corresponding equivalent in any particular non-web tech).